cmd/tether/attach.go

```diff
@@ -140,78 +171,121 @@ func (t *attachServerSSH) start() error {
 	}
 	t.sshConfig.AddHostKey(pkey)
 
-	t.enabled = true
+	t.Enabled()
 	go t.run()
 
 	return nil
 }
 
 // stop is not thread safe with start
-func (t *attachServerSSH) stop() {
+func (t *attachServerSSH) stop() error {
 	defer trace.End(trace.Begin("stop attach server"))
 
-	if t == nil || !t.enabled {
-		return
+	if t == nil {
+		err := fmt.Errorf("attach server is not configured")
+		log.Error(err)
+		return err
+	}
+
+	if !t.IsEnabled() {
+		err := fmt.Errorf("attach server is not enabled")
+		log.Error(err)
+		return err
 	}
 
-	t.enabled = false
-	conn := t.conn
-	t.conn = nil
+	log.Debugf("Setting enabled to false")
+	t.Disabled()
 
-	if conn != nil {
-		(*conn).Close()
+	log.Debugf("Canceling the context")
+	t.cancel()
+
+	log.Debugf("Acquiring the connection lock")
+	t.conn.Lock()
+	if t.conn.conn != nil {
```

Comment:
Correct me if I'm wrong but what you are saying is only valid (I think) if you are calling stop and start concurrently, right? Those are not exported, instead we have Start and Stop which syncronized via mutex so this shouldn't be a problem outside of this package. And if you are calling start/stop within the package then I think correct answer could be "deal with it" :)